File: src/main/java/net/shadowfacts/craftingslabs/multipart/crafting/PartCraftingSlab.java
package net.shadowfacts.craftingslabs.multipart.crafting;

import lombok.Getter;
import mcmultipart.MCMultiPartMod;
import mcmultipart.multipart.*;
import mcmultipart.raytrace.PartMOP;
import net.minecraft.block.BlockSlab;
import net.minecraft.block.BlockSlab.EnumBlockHalf;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyEnum;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.shadowfacts.craftingslabs.CraftingSlabs;

import java.util.Arrays;
import java.util.EnumSet;
import java.util.List;

/**
 * @author shadowfacts
 */
public class PartCraftingSlab extends Multipart implements ISlottedPart {

	public static final PropertyEnum<BlockSlab.EnumBlockHalf> HALF = PropertyEnum.create("half", BlockSlab.EnumBlockHalf.class);

	@Getter
	private EnumBlockHalf half = EnumBlockHalf.BOTTOM;


	public PartCraftingSlab() {
	}

	public PartCraftingSlab(EnumBlockHalf half) {
		this.half = half;
	}

	private AxisAlignedBB getBoundingBox() {
		if (half == EnumBlockHalf.BOTTOM) {
			return new AxisAlignedBB(0, 0, 0, 1, .5, 1);
		} else {
			return new AxisAlignedBB(0, .5, 0, 1, 1, 1);
		}
	}

	@Override
	public void addSelectionBoxes(List<AxisAlignedBB> list) {
		list.add(getBoundingBox());
	}

	@Override
	public void addCollisionBoxes(AxisAlignedBB mask, List<AxisAlignedBB> list, Entity collidingEntity) {
		AxisAlignedBB box = getBoundingBox();
		if (box.intersectsWith(mask)) {
			list.add(box);
		}
	}

	@Override
	public ItemStack getPickBlock(EntityPlayer player, PartMOP hit) {
		return new ItemStack(CraftingSlabs.items.craftingSlab);
	}

	@Override
	public List<ItemStack> getDrops() {
		return Arrays.asList(new ItemStack(CraftingSlabs.items.craftingSlab));
	}

	@Override
	public Material getMaterial() {
		return Material.WOOD;
	}

	public boolean onActivated(EntityPlayer player, EnumHand hand, ItemStack heldItem, PartMOP hit) {
		player.openGui(CraftingSlabs.instance, 2, player.worldObj, hit.getBlockPos().getX(), hit.getBlockPos().getY(), hit.getBlockPos().getZ());
		return true;
	}

	@Override
	public void writeToNBT(NBTTagCompound tag) {
		tag.setBoolean("Half", half == EnumBlockHalf.BOTTOM);
	}

	@Override
	public void readFromNBT(NBTTagCompound tag) {
		half = tag.getBoolean("Half") ? EnumBlockHalf.BOTTOM : EnumBlockHalf.TOP;
	}

	@Override
	public void readUpdatePacket(PacketBuffer buf) {
		half = buf.readBoolean() ? EnumBlockHalf.BOTTOM : EnumBlockHalf.TOP;
	}

	@Override
	public void writeUpdatePacket(PacketBuffer buf) {
		buf.writeBoolean(half == EnumBlockHalf.BOTTOM);
	}

	@Override
	public BlockStateContainer createBlockState() {
		return new BlockStateContainer(MCMultiPartMod.multipart, HALF);
	}

	@Override
	public IBlockState getExtendedState(IBlockState state) {
		return state.withProperty(HALF, half);
	}

	@Override
	public ResourceLocation getModelPath() {
		return new ResourceLocation("craftingslabs", "partCraftingSlab");
	}

	public void setHalf(EnumBlockHalf half) {
		this.half = half;
	}

	@Override
	public EnumSet<PartSlot> getSlotMask() {
		switch (half) {
			case BOTTOM:
				return EnumSet.of(PartSlot.DOWN);
			case TOP:
				return EnumSet.of(PartSlot.UP);
		}
		return EnumSet.noneOf(PartSlot.class);
	}

	@Override
	public float getHardness(PartMOP hit) {
		return 0.3f;
	}

	@Override
	public boolean isToolEffective(String type, int level) {
		return type.equals("axe");
	}

	public static PartCraftingSlab getCraftingSlab(World world, BlockPos pos, PartSlot slot) {
		IMultipartContainer container = MultipartHelper.getPartContainer(world, pos);
		if (container == null) {
			return null;
		}

		if (slot != null) {
			ISlottedPart part = container.getPartInSlot(slot);
			if (part instanceof PartCraftingSlab) {
				return ((PartCraftingSlab)part);
			}
		} else {
			ISlottedPart down = container.getPartInSlot(PartSlot.DOWN);
			if (down instanceof PartCraftingSlab) {
				return (PartCraftingSlab)down;
			} else {
				ISlottedPart up = container.getPartInSlot(PartSlot.UP);
				if (up instanceof PartCraftingSlab) {
					return (PartCraftingSlab)up;
				}
			}
		}

		return null;
	}

}

File: src/main/kotlin/net/shadowfacts/craftingslabs/multipart/PartCraftingSlab.kt
package net.shadowfacts.craftingslabs.multipart

import mcmultipart.MCMultiPartMod
import mcmultipart.multipart.ISlottedPart
import mcmultipart.multipart.Multipart
import mcmultipart.multipart.MultipartHelper
import mcmultipart.multipart.PartSlot
import mcmultipart.raytrace.PartMOP
import net.minecraft.block.BlockSlab
import net.minecraft.block.material.Material
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.wrapper.InvWrapper
import net.shadowfacts.craftingslabs.CraftingSlabs
import net.shadowfacts.craftingslabs.MODID
import net.shadowfacts.craftingslabs.gui.GUIHandler
import net.shadowfacts.craftingslabs.util.InventoryCrafting
import java.util.*

/**
 * @author shadowfacts
 */
class PartCraftingSlab(var half: BlockSlab.EnumBlockHalf): Multipart(), ISlottedPart {

	companion object {
		val HALF: PropertyEnum<BlockSlab.EnumBlockHalf> = PropertyEnum.create("half", BlockSlab.EnumBlockHalf::class.java)

		fun getCraftingSlab(world: World, pos: BlockPos, slot: PartSlot?): PartCraftingSlab? {
			val container = MultipartHelper.getPartContainer(world, pos) ?: return null

			if (slot != null) {
				val part = container.getPartInSlot(slot)
				if (part is PartCraftingSlab) {
					return part
				}
			} else {
				val down = container.getPartInSlot(PartSlot.DOWN)
				if (down is PartCraftingSlab) {
					return down
				} else {
					val up = container.getPartInSlot(PartSlot.UP)
					if (up is PartCraftingSlab) {
						return up
					}
				}
			}

			return null
		}
	}

	val inventory = InventoryCrafting(3, 3)
	val wrapper = InvWrapper(inventory)

	constructor(): this(BlockSlab.EnumBlockHalf.BOTTOM)

	private fun getBoundingBox(): AxisAlignedBB {
		return when (half) {
			BlockSlab.EnumBlockHalf.BOTTOM -> AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 0.5, 1.0)
			BlockSlab.EnumBlockHalf.TOP -> AxisAlignedBB(0.0, 0.5, 0.0, 1.0, 1.0, 1.0)
		}
	}

	override fun addSelectionBoxes(list: MutableList<AxisAlignedBB>) {
		list.add(getBoundingBox())
	}

	override fun addCollisionBoxes(mask: AxisAlignedBB, list: MutableList<AxisAlignedBB>, collidingEntity: Entity?) {
		val box = getBoundingBox()
		if (box.intersectsWith(mask)) {
			list.add(box)
		}
	}

	override fun getPickBlock(player: EntityPlayer?, hit: PartMOP?): ItemStack {
		return ItemStack(CraftingSlabs.items.craftingSlab)
	}

	override fun getDrops(): MutableList<ItemStack>? {
		return mutableListOf(ItemStack(CraftingSlabs.items.craftingSlab))
	}

	override fun getMaterial(): Material {
		return Material.WOOD
	}

	override fun onActivated(player: EntityPlayer, hand: EnumHand, heldItem: ItemStack?, hit: PartMOP): Boolean {
		val id = when (half) {
			BlockSlab.EnumBlockHalf.BOTTOM -> GUIHandler.CRAFTING_BOTTOM
			BlockSlab.EnumBlockHalf.TOP -> GUIHandler.CRAFTING_TOP
		}
		player.openGui(CraftingSlabs, id, player.world, hit.blockPos.x, hit.blockPos.y, hit.blockPos.z)
		return true
	}

	override fun writeToNBT(tag: NBTTagCompound): NBTTagCompound {
		tag.setBoolean("half", half == BlockSlab.EnumBlockHalf.BOTTOM)
		inventory.writeToNBT(tag)
		return tag
	}

	override fun readFromNBT(tag: NBTTagCompound) {
		half = if (tag.getBoolean("half")) BlockSlab.EnumBlockHalf.BOTTOM else BlockSlab.EnumBlockHalf.TOP
		inventory.readFromNBT(tag)
	}

	override fun writeUpdatePacket(buf: PacketBuffer) {
		buf.writeBoolean(half == BlockSlab.EnumBlockHalf.BOTTOM)
	}

	override fun readUpdatePacket(buf: PacketBuffer) {
		half = if (buf.readBoolean()) BlockSlab.EnumBlockHalf.BOTTOM else BlockSlab.EnumBlockHalf.TOP
	}

	override fun createBlockState(): BlockStateContainer {
		return BlockStateContainer(MCMultiPartMod.multipart, HALF)
	}

	override fun getActualState(state: IBlockState, world: IBlockAccess, pos: BlockPos): IBlockState {
		return state.withProperty(HALF, half)
	}

	override fun getModelPath(): ResourceLocation {
		return ResourceLocation(MODID, "partCraftingSlab")
	}

	override fun getSlotMask(): EnumSet<PartSlot> {
		return when (half) {
			BlockSlab.EnumBlockHalf.BOTTOM -> EnumSet.of(PartSlot.DOWN)
			BlockSlab.EnumBlockHalf.TOP -> EnumSet.of(PartSlot.UP)
		}
	}

	override fun getHardness(hit: PartMOP?): Float {
		return 0.3f
	}

	@Deprecated("")
	override fun isToolEffective(type: String, level: Int): Boolean {
		return type == "axe"
	}

	override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
		return capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY
	}

	override fun <T: Any?> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
		return if (capability == CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) wrapper as T else null
	}

	public override fun markDirty() {
		super.markDirty()
	}

}